### PR TITLE
creating new user now sends email

### PR DESCRIPTION
**Pull request by LeekSwan**
## Email service for new users and added friends
- Sends email to new user on Login
- Sends email to added friend using friend email

### TODO: 
- create unique email domain for social circle. Maybe create a social circle acc?

### CODE REVIEW

File: services/mail.js [hunk 1]
```diff
@@ -11,6 +12,30 @@ async function sendTestEmailToMyself (myPersonalEmail) {
   return mailClient.send(msg)
 }
 
+async function sendNewUserEmail (data) {
+  const msg = {
+    to: data.email,
+    from: process.env.EMAIL,
```
Comment (jasontyu) [inline services/mail.js:18]: [nit] `process.env.EMAIL` should be more descriptive (maybe `EMAIL_SENDER_ADDRESS` or something)

File: mail/emailTemplate.js [hunk 2]
```diff
@@ -0,0 +1,241 @@
+
```
Comment (jasontyu) [inline mail/emailTemplate.js:1]: [nit] make this file plural

Comment (jasontyu) [reply mail/emailTemplate.js:1]: Your implementation is pretty smart

File: routes/index.js [hunk 3]
```diff
@@ -15,9 +15,9 @@ router.get('/test-db', async function (req, res) {
 })
 
 router.get('/test-mail', function (req, res) {
-  const myEmail = ''
-  const sendEmail = false
-  // const sendEmail = true
+  const myEmail = process.env.EMAIL
+  // const sendEmail = false
+  const sendEmail = true
```
Comment (jasontyu) [inline routes/index.js:20]: [nit] Undo these changes (they should be temporary / local only)

File: services/mail.js [hunk 4]
```diff
@@ -1,4 +1,5 @@
 const mailClient = require('../mail/sendgrid.js')
+const newEmail = require('../mail/emailTemplate.js')
```
Comment (jasontyu) [inline services/mail.js:2]: [nit] it's not clear that `newEmail` is a class with 2 functions. I think `emailTemplate.newUser` makes slightly more sense than `newEmail.newUser`. Even better would be to rename the functions to `buildNewUserEmail` or something, and you can just import the function directly `const { buildNewUserEmail } = require('..')`

File: services/users.js [hunk 5]
```diff
@@ -25,17 +26,16 @@ async function login (secret) {
 }
 
 // TODO: refactor this so it makes more sense
-async function addFriend (userId, friendFName, friendLName, friendEmail) {
+async function addFriend (userId, firstName, lastName, friendFName, friendLName, friendEmail) {
```
Comment (jasontyu) [inline services/users.js:29]: Try to avoid adding arguments  to functions if you can help it. In this case, I think it's better to get user's firstName and lastName from the database.

File: routes/api.js [hunk 6]
```diff
@@ -49,8 +49,8 @@ api.get('/user/:secret/exposure', async function (req, res) {
 
 // TODO: check if friend added is the user itself
 api.post('/friendships', async function (req, res) {
-  const { userId, friendFName, friendLName, friendEmail } = req.body
-  UserService.addFriend(userId, friendFName, friendLName, friendEmail)
+  const { userId, firstName, lastName, friendFName, friendLName, friendEmail } = req.body
```
Comment (jasontyu) [inline routes/api.js:52]: 400 if any of the fields are missing

File: services/users.js [hunk 7]
```diff
@@ -15,7 +16,7 @@ async function signup (firstName, lastName, email) {
   }
   const secret = uuidv4()
   UserModel.create(firstName, lastName, email, secret)
```
Comment (jasontyu) [inline services/users.js:18]: [nit] unrelated, but probably should `await UserModel.create` before returning to user, otherwise they might try to add a friend before their account exists.

File: services/users.js (hunk 5 above)
Comment (LeekSwan) [reply services/users.js:29]: Since we already have the user data, shouldnt we limit the amount of operations we make to the db? Or does it not matter much?

Comment (jasontyu) [reply services/users.js:29]: Gotta balance performance with maintainability. I'm not sure how long it takes, but it's probably quite fast. If this causes a bug or makes it annoying to change, then it's better to leave out a non-trivial optimization to make development easier.

We want to avoid the anti-pattern called [premature optimization](https://wiki.c2.com/?PrematureOptimization).

If we start running into issues with performance, then we should start with the exposure algorithm, which hits the DB for every single friend in the network (lol).

Comment (jasontyu) [reply services/users.js:29]: Plus, the DB should be our Single Source of Truth for name info -- letting the request do it opens us up to bugs with inconsistent name info (or even attacks with malicious? name info)

Comment (LeekSwan) [reply services/users.js:29]: Ahh ok that makes sense. Thanks for clarifying.
